synapse/handlers/room.py

```diff
@@ -1450,25 +1451,37 @@ async def get_event_for_timestamp(
                         remote_response,
                     )
 
-                    # TODO: Do we want to persist this as an extremity?
-                    # TODO: I think ideally, we would try to backfill from
-                    # this event and run this whole
-                    # `get_event_for_timestamp` function again to make sure
-                    # they didn't give us an event from their gappy history.
                     remote_event_id = remote_response.event_id
                     origin_server_ts = remote_response.origin_server_ts
 
+                    # Backfill this event so we can get a pagination token for
+                    # it with `/context` and paginate `/messages` from this
+                    # point.
+                    #
+                    # FIXME: After this backfill, we might want to run this
+                    # whole `get_event_for_timestamp` function again to make
+                    # sure they didn't give us an event from their gappy
+                    # history. Also need a heuristic for when to stop recursing
+                    # if they keep giving us gappy results.
```

Comment:
On a client, `GET /timestamp_to_event?dir=b&ts=<the_ts>`, the ideal answer is `event4`

Actual events:
```
create_event
event1
event2
event3
event4
<ts corresponds to here>
event5
event6
event7
event8
event9
```


On our local server, we find the `create_event` locally next to a forward gap, so we go ask the remote homeserver.
Local server:
```
create_event
<gap>
event9
```

On the remote server we ask, they find `event2` is closest which is returned. But it's also next to a forward gap and not the true closest event.
Remote server:
```
create_event
event1
event2
<gap>
event9
```

So on our local server, after we backfill `event2`, we can see that it's still next to a forward gap. Currently, we just return the closest event (the local one or the remote one we asked about).

In the future, it might be good to be smarter about figuring out the gap to see if we can determine the true closest (assuming good intent in the DAG to event `origin_server_ts` relationship which can be made up). Maybe we ask another homeserver and see if we get the same answer. We could look forward and backfill to try to close the gap, etc. Maybe ideally, the federation `/timestamp_to_event` endpoint returns events on both sides of the timestamp to reconcile the gap in less steps.

Reply:
right, yup, I'm with you now, but I don't think the comment explains this very clearly - or at least, it's not the terminology I would use.

How about:

```
                    # TODO: the requested timestamp may lie in a part of
                    #   the event graph that the remote server *also* doesn't have,
                    #   in which case they will have returned another event which
                    #   may be nowhere near the requested timestamp.
                    #   In future, we may need to ask other homeservers, and/or
                    #   extend `/timestamp_to_event` to return events on *both*
                    #   sides of the timestamp to help reconcile the gap.
```

(by the way, note that there is a loose convention that the second and subsequent lines of a multiline FIXME or TODO should be indented, though I don't know if this is honoured by anything other than the JetBrains IDEs).

Extensions to `/timestamp_to_event` certainly seem like they are going to be needed here.

Reply:
Updated with a slightly modified version of that 👍